iron-spi-aws/src/main/java/io/axway/iron/spi/aws/AwsUtils.java

```diff
@@ -36,29 +45,54 @@ public static void setAws(AwsClientBuilder builder, //
      * @param action the action to retry
      * @param retryLimit retry number limit
      * @param durationInMillis duration between each retry
-     * @throws LimitExceededException after retry exhausted
+     * @throws AwsException after retry exhausted
      */
     public static <T> T performAmazonActionWithRetry(String actionLabel, Supplier<T> action, int retryLimit, int durationInMillis) {
         int retryCount = 0;
         do {
             try {
                 return action.get();
-            } catch (LimitExceededException lee) {
+            } catch (LimitExceededException | ProvisionedThroughputExceededException | KMSThrottlingException e) {
                 // We should just wait a little time before trying again
                 int remainingRetries = retryLimit - retryCount;
-                LOG.debug("LimitExceededException caught", args -> args.add("action", actionLabel).add("remainingRetryCount", remainingRetries));
-            }
-            try {
-                LOG.debug("Throttling", args -> args.add("action", actionLabel).add("durationMs", durationInMillis));
-                Thread.sleep(durationInMillis);
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new AwsException("Thread interrupted while performing action", args -> args.add("action", actionLabel), e);
+                LOG.debug("Amazon exception caught",
+                          args -> args.add("exception", e.getClass().getName()).add("action", actionLabel).add("remainingRetryCount", remainingRetries));
             }
+            sleepUntilInterrupted(actionLabel, durationInMillis);
         } while (retryCount++ < retryLimit);
         throw new AwsException("Limit exceeded, all retries failed", args -> args.add("action", actionLabel).add("retryLimit", retryLimit));
     }
 
+    /**
+     * Try to perform an Amazon action and increase the duration between requests if some exception is exceeding resource usage exception is thrown.
+     *
+     * @param actionLabel action label used for logging purpose only
+     * @param action the action to retry
+     * @param durationBetweenRequests duration between each retry
+     */
+    public static <T> Optional<T> tryAmazonAction(String actionLabel, Supplier<T> action, AtomicLong durationBetweenRequests) {
```

Comment:
leak of encapsulation : if you need a state (eg durationBetweenRequest) i prefer to have the corresponding object rather an extra parameter